### ISSUE TITLE
fd_cleanup in branch development-2.x  contains a bug

### ISSUE THREAD

**Issue by eriksjolund** [apptainer/singularity#1447]
## Version of Singularity:

git branch
development-2.x

The code I am referring to has not yet been released in any official release, but lives in the file
src/util/util.c in the git branch development-2.x.  In for instance this commit:

https://github.com/singularityware/singularity/blob/4c386b8273eddb104d1533f9036be1b093f86990/src/util/util.c

The if statment 

    if ( is_dir(fd_path) < 0 || is_sock(fd_path) < 0 )

is always true.

I would guess, this what is meant

    if ( is_dir(fd_path) < 0 && is_sock(fd_path) < 0 )

When tried to fix that bug, I also got curious if not the whole function could be rewritten:
I tried out two approaches:

Only iterating over the entries in /proc/self/fd:
https://github.com/eriksjolund/singularity/commits/optimize_fd_cleanup/src/util/util.c

Iterating over all the file descriptors up til sysconf(_SC_OPEN_MAX)
https://github.com/eriksjolund/singularity/commits/simplify_fd_cleanup/src/util/util.c

**Comment by dtrudg** [comment]
Hi @cclerget - would you be able to take a look at this?

**Comment by cclerget** [comment]
Hi @eriksjolund, @dctrud. Yes, I wrote a fix and modified it based on your suggestions @eriksjolund. Thank you for your help @eriksjolund !

**Comment by jthiltges** [comment]
Is your fix available on github, @cclerget? I'd prefer the opendir() version above, rather than closing all possible fds.

In docker 1.12.6 (which ships with the CentOS 7 extras), the default limit for open files is 1048576. Calling sysconf(), stat(), and close() for each fd causes a significant delay in singularity 2.4.6 startup. A simple echo test inside a container took about 13 seconds to run, vs <0.1s after lowering the open file limit.

Regards,
John

```console
# Testing with the default limit of 1M open files
$ ulimit -n
1048576
$ time /usr/bin/singularity exec /cvmfs/singularity.opensciencegrid.org/opensciencegrid/osgvo-el6:latest echo "Hello World"
Hello World

real    0m13.618s
user    0m1.853s
sys     0m11.720s

# Retesting after lowering the limit
$ ulimit -n 1024
$ time /usr/bin/singularity exec /cvmfs/singularity.opensciencegrid.org/opensciencegrid/osgvo-el6:latest echo "Hello World"
Hello World

real    0m0.069s
user    0m0.021s
sys     0m0.029s
```

**Comment by cclerget** [comment]
@jthiltges Yes initial fix was bad and incomplete. Fix which will come with next minor release use opendir but I can't say exactly when that will be release

And wow ! 13 seconds is incredibly slow

**Comment by cclerget** [comment]
Hi @eriksjolund , I'll close the issue since the fix is part of release 2.5. Feel free to re-open it if you find something wrong. Thank you !
